src/org/omegat/gui/theme/IThemeInitializer.java

```diff
@@ -0,0 +1,34 @@
+/**************************************************************************
+ OmegaT - Computer Assisted Translation (CAT) tool
+          with fuzzy matching, translation memory, keyword search,
+          glossaries, and translation leveraging into updated projects.
+
+ Copyright (C) 2021 Hiroshi Miura
+               Home page: http://www.omegat.org/
+               Support center: https://omegat.org/support
+
+ This file is part of OmegaT.
+
+ OmegaT is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ OmegaT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ **************************************************************************/
+package org.omegat.gui.theme;
+
+public interface IThemeInitializer {
```

Comment:
Current interface is to pass a argument for LookAndFeelInfo class constructor and setup() method.
It is natural both class has same methods, getName() and getClassName().

Why not asking theme plugins to provide an object that extends LookAndFeelInfo?
We can provide a class `public class ThemeInitializer extends LookAndFeelInfo` that have `void setup()` as a extension point.

Reply:
We can change `registerTheme` just accept `LookAndFeelInfo` class.

```
public static void registerTheme(UIManager.LookAndFeelInfo info) {
         UIManager.installLookAndFeel(info);
         THEME_KEY_SET.add(info.toString());
```

Then Default theme becomes like

```
    public static void loadPlugins() {
        UIDesignManager.registerTheme(new DefaultFlatThemeLaf());
    }

    public static class DefaultFlatThemeLaf extends UIManager.LookAndFeelInfo {

        public DefaultFlatThemeLaf() {
            super(Preferences.THEME_DEFAULT, defaultLaf);
            installFlatDesign();
        }
```

This brings simplicity for API. And also we can guarantee that a LaF class has a single design.
When developer want to customize LaF library,  simply inherit it and extend it in constructor as same as default theme.

Then we can use class name as a configuration key. 

Is it make sense?

Reply:
After some experiment, I want to forget the idea here 🧠

Reply:
> This brings simplicity for API. And also we can guarantee that a LaF class has a single design.
> When developer want to customize LaF library, simply inherit it and extend it in constructor as same as default theme.

I think this is a great idea. What was the problem with it?

I don't think it makes a lot of sense to have separate themes that require configuration outside of the class itself.

Reply:
```
    public static class DefaultFlatThemeLaf extends UIManager.LookAndFeelInfo {

        public DefaultFlatThemeLaf() {
            super(Preferences.THEME_DEFAULT, defaultLaf);
            installFlatDesign();
        }
```
A class inherit from  **LookAndFeelInfo** not **LookAndFeel** class, but I try to run `installFlatDesign()` in constructor.
It does not work.

`UIManager.put9)` should called after LaF class is registered with `UIManager.setLookAndFeel()` called.

Above code does it in constructor of  child of `LookAndFeelInfo`. It is wrong.

Reply:
Thanks.

# Terminology

First, I think we should clarify some things by defining terms:

- A "LAF" is an implementation of [`LookAndFeel`](https://docs.oracle.com/javase/7/docs/api/javax/swing/LookAndFeel.html)
- A "configuration" is some values set to `UIManager` such as in `DefaultFlatTheme#installFlatDesign()` 
- A "theme" is a combination of 1) LAF and 2) configuration

# LAF and configuration

I note that `LookAndFeel` has a method `getDefaults` that returns an instance of `UIDefaults`. I believe that this is how a LAF is supposed to provide its "configuration".

# Default theme

The default theme doesn't tie itself to a particular LAF; it uses the default LAF, which depends on the system.

The default theme also has a lot of associated configuration, which we install as a secondary step (`DefaultFlatTheme#installFlatDesign()`).

The secondary step is only necessary because the LAF is not known until runtime; otherwise we could create a `LookAndFeel` class that inherits from some existing LAF and provide our configuration via `getDefaults`.

(Actually I have an idea for how to do this: create a `LookAndFeel` class that merely wraps the system LAF. I made this locally but I couldn't test it due to this error: https://github.com/omegat-org/omegat/pull/100#discussion_r646132685)

# Default configuration

When I tried your test plugin, I found the result quite ugly on macOS because `installFlatDesign` was not called. I think we might need to install at least some of this configuration by default, regardless of the theme. Or we can modularize it and make it easy for themes to adapt; see below.

# Theme plugins

I imagine that theme plugins will want to supply a LAF and configuration. I have a hard time imagining a theme plugin that would supply *only* a configuration.

If the plugin supplies a LAF, per above I believe they can supply their configuration as well from `getDefaults`.

If the plugin *only* supplies configuration, I believe it can already do so with an `IApplicationEventListener` that implements `onApplicationStartup()` (I haven't tested this; of course there are severe limitations such as the theme will always be activated if the plugin is loaded.)

# Improving the configuration

I think we could package our configuration (what happens in `installFlatDesign`) in an instance of `UIDefaults`, or make it take action on an instance of `UIDefaults` instead of directly on `UIManager`.

This would let theme plugins easily reuse (with modifications if desired) the default configuration.

What do you think?

Reply:
Sorry, I should be clear about action items:

1. Please give me your opinion on my points above
2. Please see if you can reproduce (and fix?) the error I mentioned in https://github.com/omegat-org/omegat/pull/100#discussion_r646132685
3. Once the error is resolved, I will test my implementation of the default theme entirely encapsulated in a LAF. If this works, we can discard the entire concept of "theme initializers"

Reply:
1. It is better to realize your idea above.  I just failed the last trial because of my misunderstandings.
2. I've fixed the issue [No.2](https://github.com/omegat-org/omegat/pull/100#discussion_r646132685)
3. I don't have an instant idea how to realize an "encapsulating" class. I'd like to wait @amake for No.3.

Reply:
@amake I believe I've realized a LaF class implementation of the default theme entirely encapsulated in a LaF now.
Sample plugin project is also implemented as same way.
https://github.com/miurahr/omegat-flatlaf/tree/75ec61c4c8699ead4b89c48bf2d2c15ea6971c3f

Reply:
Thanks, and sorry I haven't been able to work on this for the last few days. I hope to get back to it this weekend.

Your implementation is mostly in line with what I was working on; just a few details are different. I'll give it a closer look soon.